vnext/JSI/Shared/ChakraObjectRef.h

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "jsi/jsi.h"
+
+#ifdef CHAKRACORE
+#include "ChakraCore.h"
+#else
+#ifndef USE_EDGEMODE_JSRT
+#define USE_EDGEMODE_JSRT
+#endif
+#include <jsrt.h>
+#endif
+
+#include <exception>
+#include <string>
+
+namespace Microsoft::JSI {
+
+inline void CrashUponChakraError(JsErrorCode error) {
+  if (error != JsNoError) {
+    std::terminate();
+  }
+}
+
+void ThrowUponChakraError(
+    JsErrorCode error,
+    const char *const chakraApiName = nullptr);
+
+// An shared_ptr like RAII Wrapper for JsRefs, which are references to objects
+// owned by the garbage collector. These include JsContextRef, JsValueRef, and
+// JsPropertyIdRef, etc. ChakraObjectRef ensures that JsAddRef and JsRelease
+// are called upon initialization and invalidation, respectively. It also allows
+// users to implicitly convert it into a JsRef. A ChakraObjectRef must only be
+// initialized once and invalidated once.
+class ChakraObjectRef {
+ public:
+  ChakraObjectRef() noexcept {}
+  inline explicit ChakraObjectRef(JsRef ref) {
```

Comment:
>) [](start = 43, length = 1)

(nit) noexcept?

Reply:
Unfortunately, during the call to Initialize, we throw if JsAddRef fails. That is why I did not mark this function as noexcept.

---
In reply to: [336144142](https://github.com/microsoft/react-native-windows/pull/3445#discussion_r336144142) [](ancestors = 336144142)

Reply:
From discussion. Leave as is for this CR. 

Let's track it separately with work item:
[https://office.visualstudio.com/OC/_workitems/edit/3712909/](https://office.visualstudio.com/OC/_workitems/edit/3712909/)

---
In reply to: [336146103](https://github.com/microsoft/react-native-windows/pull/3445#discussion_r336146103) [](ancestors = 336146103,336144142)